src/core/trulens/core/utils/python.py

```diff
@@ -1148,118 +1150,31 @@ def rewrap(obj_):
 
 
 # Class utilities
-
-
-@dataclasses.dataclass
-class SingletonInfo(Generic[T]):
-    """
-    Information about a singleton instance.
+class SingletonPerName(type):
     """
-
-    val: T
-    """The singleton instance."""
-
-    cls: Type[T]
-    """The class of the singleton instance."""
-
-    frameinfo_codeline: Optional[str]
-    """The frame where the singleton was created.
-
-    This is used for showing "already created" warnings. This is intentionally
-    not the frame itself but a rendering of it to avoid maintaining references
-    to frames and all of the things a frame holds onto.
-    """
-
-    name: Optional[str] = None
-    """The name of the singleton instance.
-
-    This is used for the SingletonPerName mechanism to have a separate singleton
-    for each unique name (and class).
-    """
-
-    def __init__(self, name: str, val: Any):
-        self.val = val
-        self.cls = val.__class__
-        self.name = name
-        self.frameinfo_codeline = code_line(
-            caller_frameinfo(offset=2), show_source=True
-        )
-
-    def warning(self):
```

Comment:
Can the warning mechanism be integrated into meta `__call__`? Track the frame for debugging when singleton created and then add a special method for singleton children that checks whether a warning should be issued:

```python
def init_ignored(cls, frame, *args, **kwargs):
     if # some args are not none:
         print("Singleton already created for class .... # rest of message)
```